scheduler/scheduler-node/src/main/java/org/ow2/proactive/scheduler/task/ProgressFileReaderPoller.java

```diff
@@ -0,0 +1,222 @@
+ * Contact: contact@activeeon.com
+ *
+ * This library is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Affero General Public License
+ * as published by the Free Software Foundation: version 3 of
+ * the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive.scheduler.task;
+
+import static com.google.common.io.Files.*;
+import static org.apache.commons.io.FileUtils.deleteQuietly;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileTime;
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+
+import org.apache.log4j.Logger;
+import org.ow2.proactive.scheduler.common.task.TaskId;
+import org.ow2.proactive.scheduler.core.properties.PASchedulerProperties;
+import org.ow2.proactive.scheduler.task.utils.ForkerUtils;
+
+
+/**
+ * ProgressFileReader is in charge of:
+ * - creating a progress file in the specified working directory
+ * - detecting changes on a progress file efficiently
+ * - reading new value and saving it in memory
+ * - exposing the last read value
+ * <p>
+ * Instances of this class are NOT thread-safe.
+ *
+ * @author The ProActive Team
+ */
+public class ProgressFileReaderPoller {
+
+    private static final Logger logger = Logger.getLogger(ProgressFileReaderPoller.class);
+
+    private static final String PROGRESS_FILE_DIR = ".tasks-progress";
+
+    private Path progressFileDir;
+
+    private Path progressFile;
+
+    private volatile int progress;
+
+    private Thread watchDogThread;
+
+    private Set<Listener> observers;
+
+    public ProgressFileReaderPoller() {
+        // mainly for test purposes
+        observers = new HashSet<>(0);
+    }
+
+    public boolean start(File workingDir, TaskId taskId) {
+        String progressFileName = "job-" + taskId.getJobId().value() + "-task-" + taskId.value() + "-" +
+                                  UUID.randomUUID() + ".progress";
+
+        return start(workingDir, progressFileName);
+    }
+
+    boolean start(File workingDir, String filename) {
+        try {
+            createProgressFile(workingDir, filename);
+
+            watchDogThread = new Thread(new ProgressFileReaderThread(Duration.ofMillis(PASchedulerProperties.SCHEDULER_PROGRESS_FILE_READER_POLL_TIMEOUT.getValueAsInt())));
```

Comment:
you cannot use PASchedulerProperties in this context as this code is executed on the node.

Reply:
why not use ScheduledExecutorService.scheduleAtFixedRate instead of Thread ?